crates/sui-core/src/event_handler.rs

```diff
@@ -20,30 +22,63 @@ use crate::streamer::Streamer;
 
 pub const EVENT_DISPATCH_BUFFER_SIZE: usize = 1000;
 
-pub struct EventHandler {
+pub struct EventHandler<ES: EventStore> {
     module_cache: SyncModuleCache<ResolverWrapper<AuthorityStore>>,
     event_streamer: Streamer<EventEnvelope, EventFilter>,
+    pub(crate) event_store: Arc<ES>,
 }
 
-impl EventHandler {
-    pub fn new(validator_store: Arc<AuthorityStore>) -> Self {
+impl<ES: EventStore> EventHandler<ES> {
+    pub fn new(validator_store: Arc<AuthorityStore>, event_store: Arc<ES>) -> Self {
         let streamer = Streamer::spawn(EVENT_DISPATCH_BUFFER_SIZE);
         Self {
             module_cache: SyncModuleCache::new(ResolverWrapper(validator_store)),
             event_streamer: streamer,
+            event_store,
         }
     }
 
-    pub async fn process_events(&self, effects: &TransactionEffects, timestamp_ms: u64) {
+    // TODO: feed in current checkpoint number
+    pub async fn process_events(
+        &self,
+        effects: &TransactionEffects,
+        timestamp_ms: u64,
+        checkpoint_num: u64,
+    ) -> SuiResult {
         // serially dispatch event processing to honor events' orders.
-        for event in &effects.events {
-            if let Err(e) = self.process_event(event, timestamp_ms).await {
+        let res: Result<Vec<_>, _> = effects
+            .events
+            .iter()
+            .map(|e| self.create_envelope(e, effects.transaction_digest, timestamp_ms))
+            .collect();
+        let envelopes = res?;
+
+        // Ingest all envelopes together at once (for efficiency) into Event Store
+        // It's good to ingest into store first before sending so that any failures in sending could
+        // use the store as a backing for reliability
```

Comment:
I don't object to storing and then sending, but "so that any failures in sending could use the store as a backing for reliability" may not be true - the subscription only deals with "real-time" data and shouldn't look into the store

Reply:
It might not be true today but might be true in the future.  Anyways not a big deal, I'll remove the last two lines.